Kernel/FileSystem/FileBackedFileSystem.cpp

```diff
@@ -107,115 +107,139 @@ class DiskCache {
     }
 
 private:
-    DiskBackedFS& m_fs;
+    FileBackedFS& m_fs;
     size_t m_entry_count { 10000 };
     KBuffer m_cached_block_data;
     KBuffer m_entries;
     bool m_dirty { false };
 };
 
-DiskBackedFS::DiskBackedFS(BlockDevice& device)
-    : m_device(device)
+FileBackedFS::FileBackedFS(File& file)
+    : m_file(file)
+    , m_file_description(FileDescription::create(*m_file))
+{
+}
+FileBackedFS::FileBackedFS(NonnullRefPtr<FileDescription> file_description)
+    : m_file(file_description->file())
+    , m_file_description(file_description)
 {
 }
 
-DiskBackedFS::~DiskBackedFS()
+FileBackedFS::FileBackedFS(NonnullRefPtr<File> file)
+    : m_file(file)
+    , m_file_description(FileDescription::create(*m_file))
 {
 }
 
-bool DiskBackedFS::write_block(unsigned index, const u8* data, FileDescription* description)
+FileBackedFS::~FileBackedFS()
 {
-#ifdef DBFS_DEBUG
-    klog() << "DiskBackedFileSystem::write_block " << index << ", size=" << data.size();
+}
+
+bool FileBackedFS::write_block(unsigned index, const u8* data, FileDescription* description, bool use_logical_block_size)
```

Comment:
Is `file_description` the parameter ever used?

Reply:
Ah, I see it is, for `allow_cache`

Reply:
A question though: how should the callers decide whether they should `use_logical_block_size` or not?

Reply:
The question still stands.

Reply:
Another bump here, too

Reply:
As I mentioned in the IRC channel, we need to distinguish between "virtual disk reads" and "file system reads". Previously, when we read the ext2 superblock, we called `device().read_raw()` to fetch the superblock. Now we can't use that call anymore because we are not backed by a `BlockDevice`, so the block size for reading the superblock is still 512 bytes like when we used `device().read_raw()`, but we need to do implement in the `FileBackedFileSystem` layer instead. 

Not only that this is necessary for fetching "pure blocks" from the disk, it will allow us to easily implement reading & writing from other types of media, like CD-ROM (with 2048 bytes per sector), or for new modern harddrives (with 4K sectors). 

In addition to that flexibility, some harddrives today have 4k sectors internally (in ATA IDENTIFY structure, the physical sector size value is 4096 bytes), and the logical sector size is 512 bytes, to stay compatible with old software that assumes 512 bytes per sector for all harddrives. It's important to know what is the physical sector size, to allow writing on 4K boundaries for optimization purposes.
You can read more about that here:
https://ata.wiki.kernel.org/index.php/ATA_4_KiB_sector_issues